### ISSUE TITLE
Schnittstelle für ONYX Aufgaben und GUI Diskussion

### ISSUE THREAD

**Issue by gelbeforelle** [gelbeforelle/SWE-SS21#6]
Letzter essentieller Planungsschritt wird die Übergabe von Klassen zur Speicherung von Tests an die Ausgabe-Elemente umfassen.

Für das Erstellen eines ONYX Tests brauche ich folgende Eigenschaften:
* Art des Aufgabe (Multiple Choice, Lückenaufgabe, Freitext)
* Kardinalität der Antwort (nur MC) - gibt an, ob mehrfache Auswahl möglich ist
* Auswertungsmuster (max./ min. mögliche Punktzahl, automatische oder manuelle Auswertung...)
* eigentlichen formatierten Text

Wir sollten uns im Klaren sein, welche Eigenschaften wir hier noch sinnvoll per Formatierung in Word erzeugen können und welche wir rein in unserem Programm abfragen.

Für die Übergabe schlage ich ein Interface "ITest" vor, von dem die Tests jeweils erben. Jeder Test sollte ein Array besitzen, welches die Namen der Antwort-Variable(n) (in ihr wird die Antwort zur späteren Auswertung gespeichert) enthält. Zudem sollten sie einen Text besitzen (Paragraph Klasse oder String[]). Hier ist zu beachten, dass Lückentexte ihre Lücken im Text kennzeichnen müssen. 

Ich denke, es wäre am einfachsten, für jede erbende Klasse von ITest nur einen Typ von Test zuzulassen und kombinierte Tests gegebenenfalls über ein ITest[] Array realisiert.

**Comment by gelbeforelle** [comment]
Ich habe mich mit Windows Forms befasst. Ich denke, wir sollten eine Farbe für Lückentext-Aufgaben reservieren. Diese Eigenschaft lässt sich einfacher über Markierungen in Word festlegen als eine GUI in C#. 

Kardinalität für Multiple Choice und automatische Bewertung lassen sich einfach über eine GUI realisieren. 

Du solltest mir bitte ein Enum übergeben, welches den Typ des Tests festlegt, sowie einen String, der den Text beinhaltet. Zu guter Letzt benötige ich einen String, der die Lösung beinhaltet, beziehungsweise ein Array, falls es sich um einen Lückentext oder Multiple Choice handelt.

**Comment by gelbeforelle** [comment]
Sieh dir mal mneinen Entwurf für eine Klasse, welche Aufgaben speichern soll, an.
Es wird ein Text gespeichert, sowie für MultipleChoice einige Aspekte zur Auswertung. Lückenaufgaben sind vorerst nicht vorgesehen, da sie eine komplexere Struktur besitzen.
Die QTI Klasse ist eine Erweiterung für StreamWriter, welche es ihnen erlaubt, im QTI Format zu schreiben.
Vielleicht könnte man auf ähnliche Weise auch den XmlParser mit der XmlReader-Klasse verbinden.

EDIT: Die MultipleChoice Aufgabe sollte natürlich ein "AnswerFormat" besitzen, keinen "AnswerKey"
![Klassendiagramm](https://i.imgur.com/Wmh6sdU.png)

**Comment by gelbeforelle** [comment]
Ich habe ein kleines Beispiel zur GUI erstellt und werde es in einem passenden Branch ablegen. Bitte beachte, dass es sich nur mit Visual Studio öffnen lässt, nicht mit Visual Studio **_Code_**.

**Comment by hannah-knst** [comment]
Ich würde vorschlagen, ich bekomme von dir ein Paragraph[]-Array übergeben..?
Das Array würde ich mit einer Schleife durchgehen.. Jeder Paragraph sollte ein Item werden, demnach gibt es max. einen Fragentypen pro Paragraph, oder?
-> So kann ich mit if-Abfragen die fStrings durchgehen, wie viele Antwortmöglichkeiten und richtige Antworten es gibt und entscheide danach, ob das ein Multiple-/Single-Choice / Question / Text ist und kann ItemParts nach der entsprechenden Unterklasse anlegen.
(Fehlen hier noch Fragentypen? Kompliziertere Sachen kann man ggf. später ergänzen…)
![image](https://user-images.githubusercontent.com/81990072/126465747-af232e63-6e28-455d-abbd-0558bc70a398.png)

Lückentexte müssten dann so implementiert werden, dass immer von einer Lücke bis zur nächsten ein Item/Paragraph erkannt wird (jeweils inklusive genau einer Lücke).

Jedes Item hat dann eine print-Methode, mit deren Hilfe ich dir Itemweise den Quelltext übergeben würde..? Wenn du das schon anders implementiert hast, kann ich dir das auch anders übergeben…

Die eigentlichen Formatierungen des Textes würde ich erstmal auf fett, kursiv, unterstrichen und Zeilenumbrüche beschränken (kommt als Info ins Tutorial).
Sämtliche Formatierungen, die dazu dienen Textabschnitte als Lücke/Antwort/… zu markieren würde ich über Farben innerhalb Word regeln. Das Ziel ist ja, den Test komplett in Word erstellen zu können, zusätzliche Abfragen in unserem Programm sind glaube zu umständlich für den Nutzer :D Dann kann man den Test eigentlich gleich über OPAL erstellen…

**Comment by gelbeforelle** [comment]
Zur Formatierung: Ich habe bereits eine QTI-Print Methode ~~auf dem Branch "qti"~~ in der "fString" Klasse angefangen, welche die für QTI formatierten Strings enthält. Sie setzt bereits alle Formatierungen, außer natürlich farblicher Hinterlegung, um. Schriftgröße ist aufgrund des stark limitierten Nutzens in Tests nicht vorgesehen.

Zur Erkennung der Farben: Bedenke, dass du einen Erkennungsalgorithmus implementieren musst, wenn du verschiedene Farbtöne unterscheiden willst. Stelle dir also die Frage, wann eine Farbe z.B. als Blau erkannt wird. Eine GUI würde ich trotzdem implementieren, alleine um die Datei auszuwählen ist sie praktisch.

Zum Klassenentwurf: Die Klasse Text ist redundant. Jedes Item besitzt bereits einen Text, daher kann er dort gespeichert werden. Die Vererbung von Choice -> Multiple, Single ist durchaus sinnvoll, allerdings halte ich Namen wie "MultipleChoice" anstatt "Multiple" und "TextQuestion" anstatt "Question" sinnvoller. Zudem sehe ich den Zweck der Kapselung von "Points" nicht, zumal diese auch als Zahl gespeichert werden könnten.

EDIT: Vielleicht solltest du auch eine Möglichkeit einbauen, zu erkennen, ob eine automatische Bewertung erwünscht ist.

**Comment by gelbeforelle** [comment]
Ich habe eine QTI-Output Klasse als Beispiel aus einer meiner "Learning by doing" Dateien erstellt. Vielleicht hilft es dir als Orientierung, welche Funktionen wir für die Tests benötigen. [Sieh es dir hier an](https://github.com/gelbeforelle/SWE-SS21/blob/qti/qti-output/Question.cs).

**Comment by gelbeforelle** [comment]
Ich habe das Diagramm in meinen Tests auf dem qti Branch etwas abgewandelt. 
Die Kapselung in `Item` habe ich entfernt. `ItemPart` wird durch `Question` als Basis abgelöst, erfüllt aber quasi die gleiche Funktion.
Multiple und Single wurden in Single-/ MultipleChoice umbenannt.
Wie erwähnt kann ich den `Link` in `Points` nicht wirklich nachvollziehen, weshalb diese nun ein double-Property sind.

Als Erkennung für AutoScoring-fähige Tests habe ich das `IScoring` Interface verwendet. Vielleicht braucht man noch eine Möglichkeit, AutoScoring bei diesen Tests zu vermeiden, falls es nicht erwünscht ist.

![Klassendiagramm](https://i.imgur.com/zzUiPJw.png)

**Comment by hannah-knst** [comment]
Ja, bei dem Text wollte ich einfach Text ohne Punktzahl, aber man kann auch einfach sämtlichen Text zur Beschreibung u.ä. mit in die Question-Klasse packen...

Den Points wollte ich einen Link zu einer Datei geben, in der man dann die Punktevergabe regelt - da gibt es wie gesagt schon einige Vorlagen auf OPAL. Wir können das aber auch erstmal auf die einfache Version beschränken, für richtige Antworten max. Punkte und für falsche Antworten 0 Punkte zu geben - in dem Fall reicht auch uint.

Bei dir verstehe ich noch nicht ganz denn Sinn des Enums Cardinality - ob das single- oder multipleChoice ist, sieht man ja daran, welche Unterklasse das ist.

**Comment by gelbeforelle** [comment]
Das ist eine gute Idee, so lässt sich die Bewertung gut kapseln. Kannst du einschätzen, ob wir den Link zeitlich noch implementieren können?
Cardinality ist erstmal recht nutzlos, da wir nur multiple und single implementieren. **Aber:** QTI erlaubt prinzipiell noch andere Kardinalitäten. Andererseits gebe ich dir recht, es war vermutlich unlogisch, eine solche Nische zu integrieren, da selbst die ausführliche Dokumentation zu QTI wenig darüber berichtet.

~~Kannst du das Enum im Quellcode entfernen? Ansonsten kann ich das übernehmen, ich habe bloß erstmal wenig Zeit.~~
EDIT: erledigt
